04.ls/ls.rb

```diff
@@ -1,19 +1,114 @@
 # frozen_string_literal: true
 
 require 'optparse'
+require 'ffi-xattr'
+require 'etc'
+
 COLUMN_COUNT = 3
 SPACE = 1
+PERMISSION_TYPE = ['--x', '-w-', '-wx', 'r--', 'r-x', 'rw-', 'rwx'].freeze
 
 def main
   options = {}
   opt = OptionParser.new
+  opt.on('-r') { |v| v }
   opt.on('-a') { |v| v }
+  opt.on('-l') { |v| v }
   opt.parse!(ARGV, into: options)
 
   path =  "#{Dir.getwd}/"
   entries = Dir.entries(path).sort
-  filtered_entries = options.key?(:a) ? entries : entries.reject { |entry| entry.start_with?('.') }
-  entries_with_suffix = append_suffix_by_file_type(filtered_entries, path)
+
+  sorted_entries = options.key?(:r) ? entries.reverse : entries
+  filtered_entries = options.key?(:a) ? sorted_entries : sorted_entries.reject { |entry| entry.start_with?('.') }
+  options.key?(:l) ? show_long_format(filtered_entries, path) : show_short_format(filtered_entries, path)
+end
+
+def show_long_format(entries, path)
+  entries_long_format = []
+  total_blocks = 0
+
+  entries.each do |entry_name|
+    entry_details = File.lstat(entry_name)
+    total_blocks += entry_details.blocks
+    entries_long_format << load_details_each_entry(entry_name, entry_details, path)
+  end
+  print_long_data(entries_long_format, total_blocks)
+end
+
+def load_details_each_entry(entry_name, entry_info, path)
+  xattr = Xattr::Lib.list(path + entry_name, @no_follow = true)
```

Comment:
```suggestion
  xattr = Xattr::Lib.list(path + entry_name, true)
```

メソッド呼び出しとしてはこれでいいと思うのですが、`@no_follow = true`みたいなコードはサンプルコードか何かがあったんでしょうか？
参考文献があればURLを教えてください〜。

Reply:
参考文献はこちらです。`@no_follow`が`true`の場合、リンクそのものの拡張属性を見ており、`false`の場合はリンクが参照しているファイルの拡張属性を見ているようです。
標準lsの場合はリンクそのものの拡張属性を見ているようだったので、今回は`true`にしました。
1) `ffi-xattr`のドキュメント： https://www.rubydoc.info/gems/ffi-xattr/0.2.0/Xattr/Lib#list-class_method
2) listxattr,llistxattrのドキュメント： https://linuxjm.osdn.jp/html/LDP_man-pages/man2/listxattr.2.html

Reply:
情報ありがとうございます。リンクを見てみましたが、`@no_follow = true`のようにインスタンス変数に代入しているコードは見当たりませんでした。
`@no_follow`はいったいどこから登場したのかな〜、というのがちょっと気になってます。
（そもそもこのメソッド呼び出しでしか登場しないので、`@no_follow`は使われない変数のはずです）

Reply:
明らかに間違った値の渡し方をしているので、`@no_follow = true`を`true`に変更しました。ありがとうございました